service/tx/service.go

```diff
@@ -182,6 +187,171 @@ func (svc *Service) CreateRawSendPool(args client.SendPoolTxRequest, reply *clie
 	return nil
 }
 
+func (svc *Service) Vote(args client.VoteRequest, reply *client.VoteReply) error {
+	if svc.evidenceStore.IsFrozenValidator(args.Address) {
+		svc.logger.Error("got allegation validator", args.Address)
+		return codes.ErrFrozenValidator
+	}
+
+	validator, err := svc.validators.Get(args.Address)
+	if err != nil {
+		svc.logger.Errorf("validator for address %s not found\n", args.Address)
+		return codes.ErrBadOwner
+	}
+
+	svc.logger.Infof("Vote transaction for address - %s\n", validator.Address)
+
+	allegation := penalization.AllegationVote{
+		Address:   validator.Address,
+		RequestID: args.RequestID,
+		Choice:    args.Choice,
+	}
+
+	data, err := allegation.Marshal()
+	if err != nil {
+		svc.logger.Error("error in serializing vote object", err)
+		return codes.ErrSerialization
+	}
+
+	uuidNew, _ := uuid.NewUUID()
+	feeAmount := svc.feeOpt.MinFee()
+
+	tx := action.RawTx{
+		Type: action.ALLEGATION_VOTE,
+		Data: data,
+		Fee: action.Fee{
+			Price: action.Amount{
+				Currency: "OLT",
+				Value:    *feeAmount.Amount,
+			},
+			Gas: 100000,
+		},
+		Memo: uuidNew.String(),
+	}
+
+	packet, err := serialize.GetSerializer(serialize.NETWORK).Serialize(tx)
+	if err != nil {
+		svc.logger.Error("error in serializing send transaction", err)
+		return codes.ErrSerialization
+	}
+
+	*reply = client.VoteReply{RawTx: packet}
+
+	return nil
+}
+
+func (svc *Service) Allegation(args client.AllegationRequest, reply *client.AllegationReply) error {
+	mv, err := svc.validators.Get(args.MaliciousAddress)
```

Comment:
I saw that the client.AllegationRequest contains an "Address" field. Does that address denote the validator who raises this allegation? Do we allow a validator to raise an allegation under somebody else's node using olclient?

Reply:
skip